File: typescript/src/key-pool.ts
/**
 * Key rotation pool for API key management.
 *
 * Round-robin key selection with dead-key tracking.
 * On 429: rotate() advances to next key.
 * On 401/403: markDead() permanently removes a key from rotation.
 * Single-threaded JS — no lock needed, just atomic index.
 *
 * Env var convention:
 *   {PROVIDER}_KEYS  — comma-separated list (preferred)
 *   {PROVIDER}_API_KEY — single key fallback (pool of 1)
 */

export class KeyPoolExhaustedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "KeyPoolExhaustedError";
  }
}

export class KeyPool {
  private readonly keys: string[];
  private readonly dead: Set<string> = new Set();
  private idx: number = 0;

  constructor(keys: string[]) {
    const cleaned = keys.map((k) => k.trim()).filter((k) => k.length > 0);
    if (cleaned.length === 0) {
      throw new Error("KeyPool requires at least one non-empty key");
    }
    this.keys = cleaned;
  }

  /** Return the current key, skipping dead keys. */
  select(): string {
    if (this.isExhausted()) {
      throw new KeyPoolExhaustedError(
        `All ${this.keys.length} keys are dead (401/403). ` +
          "Replace keys or wait for provider resolution.",
      );
    }
    for (let i = 0; i < this.keys.length; i++) {
      const key = this.keys[this.idx % this.keys.length];
      if (!this.dead.has(key)) {
        return key;
      }
      this.idx = (this.idx + 1) % this.keys.length;
    }
    // Should never reach here if isExhausted() is correct
    throw new KeyPoolExhaustedError("All keys are dead");
  }

  /** Advance to the next key (called on 429). */
  rotate(): void {
    this.idx = (this.idx + 1) % this.keys.length;
  }

  /** Mark a key as permanently failed (called on 401/403). */
  markDead(key: string): void {
    this.dead.add(key);
  }

  /** Return true if all keys are dead. */
  isExhausted(): boolean {
    return this.dead.size >= this.keys.length;
  }

  /** Number of keys still alive. */
  get aliveCount(): number {
    return this.keys.length - this.dead.size;
  }

  /** Total number of keys in the pool. */
  get totalCount(): number {
    return this.keys.length;
  }
}

/**
 * Create a KeyPool from environment variables.
 *
 * Checks {PROVIDER}_KEYS first (comma-separated), then falls back
 * to {PROVIDER}_API_KEY (pool of 1).
 *
 * @param provider - Provider name (e.g. "OPENAI", "ANTHROPIC"). Converted to uppercase.
 * @throws Error if neither env var is set.
 */
export function loadKeysFromEnv(provider: string): KeyPool {
  const providerUpper = provider.toUpperCase();

  // Try {PROVIDER}_KEYS first (comma-separated pool)
  const keysVar = `${providerUpper}_KEYS`;
  const keysRaw = process.env[keysVar] ?? "";
  if (keysRaw.trim()) {
    return new KeyPool(keysRaw.split(","));
  }

  // Fall back to {PROVIDER}_API_KEY (single key)
  const keyVar = `${providerUpper}_API_KEY`;
  const singleKey = process.env[keyVar] ?? "";
  if (singleKey.trim()) {
    return new KeyPool([singleKey]);
  }

  throw new Error(
    `No API keys found for ${providerUpper}. Set ${keysVar} (comma-separated) or ${keyVar}.`,
  );
}

File: rust/llmix-rs/src/key_pool.rs
use crate::error::KeyPoolExhaustedError;
use crate::{LlmixError, LlmixResult};
use std::collections::HashSet;
use std::env;
use std::sync::Mutex;

#[derive(Debug)]
struct KeyPoolState {
    dead: HashSet<String>,
    idx: usize,
}

#[derive(Debug)]
pub struct KeyPool {
    keys: Vec<String>,
    state: Mutex<KeyPoolState>,
}

impl KeyPool {
    pub fn new(keys: Vec<String>) -> LlmixResult<Self> {
        let mut cleaned = Vec::new();
        let mut seen = HashSet::new();

        for key in keys {
            let trimmed = key.trim();
            if trimmed.is_empty() {
                continue;
            }
            if seen.insert(trimmed.to_owned()) {
                cleaned.push(trimmed.to_owned());
            }
        }

        if cleaned.is_empty() {
            return Err(LlmixError::InvalidKeyPoolConfig(
                "KeyPool requires at least one non-empty key".to_owned(),
            ));
        }

        Ok(Self {
            keys: cleaned,
            state: Mutex::new(KeyPoolState {
                dead: HashSet::new(),
                idx: 0,
            }),
        })
    }

    pub fn select(&self) -> Result<String, KeyPoolExhaustedError> {
        let mut state = self.state.lock().expect("key pool mutex poisoned");

        if state.dead.len() >= self.keys.len() {
            return Err(KeyPoolExhaustedError {
                total_keys: self.keys.len(),
            });
        }

        for offset in 0..self.keys.len() {
            let idx = (state.idx + offset) % self.keys.len();
            let key = &self.keys[idx];
            if !state.dead.contains(key) {
                state.idx = idx;
                return Ok(key.clone());
            }
        }

        Err(KeyPoolExhaustedError {
            total_keys: self.keys.len(),
        })
    }

    pub fn rotate(&self) {
        let mut state = self.state.lock().expect("key pool mutex poisoned");

        if state.dead.len() >= self.keys.len() {
            return;
        }

        for offset in 1..=self.keys.len() {
            let idx = (state.idx + offset) % self.keys.len();
            if !state.dead.contains(&self.keys[idx]) {
                state.idx = idx;
                return;
            }
        }
    }

    pub fn mark_dead(&self, key: &str) -> LlmixResult<()> {
        if !self.keys.iter().any(|candidate| candidate == key) {
            return Err(LlmixError::UnknownKeyPoolKey(
                "Key not in pool: cannot mark unknown key as dead".to_owned(),
            ));
        }

        let mut state = self.state.lock().expect("key pool mutex poisoned");
        state.dead.insert(key.to_owned());
        Ok(())
    }

    pub fn is_exhausted(&self) -> bool {
        let state = self.state.lock().expect("key pool mutex poisoned");
        state.dead.len() >= self.keys.len()
    }

    pub fn alive_count(&self) -> usize {
        let state = self.state.lock().expect("key pool mutex poisoned");
        self.keys.len() - state.dead.len()
    }

    pub fn total_count(&self) -> usize {
        self.keys.len()
    }
}

pub fn load_keys_from_env(provider: &str) -> LlmixResult<KeyPool> {
    let provider_upper = provider
        .chars()
        .map(|character| match character {
            'a'..='z' => character.to_ascii_uppercase(),
            'A'..='Z' | '0'..='9' => character,
            _ => '_',
        })
        .collect::<String>();
    let keys_var = format!("{provider_upper}_KEYS");
    let key_var = format!("{provider_upper}_API_KEY");

    if let Ok(keys_raw) = env::var(&keys_var) {
        if !keys_raw.trim().is_empty() {
            return KeyPool::new(keys_raw.split(',').map(ToOwned::to_owned).collect());
        }
    }

    if let Ok(single_key) = env::var(&key_var) {
        if !single_key.trim().is_empty() {
            return KeyPool::new(vec![single_key]);
        }
    }

    Err(LlmixError::InvalidKeyPoolConfig(format!(
        "No API keys found for {provider_upper}. Set {keys_var} (comma-separated) or {key_var}."
    )))
}
